src/core/qgsdataprovider.h

```diff
@@ -429,6 +429,18 @@ class CORE_EXPORT QgsDataProvider : public QObject
      */
     QVariant providerProperty( int property, const QVariant &defaultValue ) const; // SIP_SKIP
 
+    /**
+     * Set whether the provider will listen to datasource notifications
+     * If set, the provider will issue notify signals.
+     *
+     * The default implementation does nothing.
+     *
+     * \see notify
+     *
+     * \since QGIS 3.0
+     */
+    virtual void setListening( bool isListening ) { Q_UNUSED( isListening ); }
```

Comment:
General though: it's better to keep such things in the cpp file. Less code in the header makes compiling faster and virtual methods cannot be inlined anyway.

Reply:
I'd like to argue (https://stackoverflow.com/questions/733737/are-inline-virtual-functions-really-a-non-sense)... just for the fun of arguing ;)